rfcs/0041-new-transaction-model.md

```diff
@@ -0,0 +1,183 @@
+  	    ; excess= local_state.excess + excess }
+  	; ledger_hash= new_ledger_hash 
+  	; fee_excess= e.fee_excess + excess_change }
+```
+
+### How this would boil down into "base" transaction SNARKs
+The idea would be to have 3 new base transaction SNARKs corresponding to the 3 forms of authentication. Each would implement the above `apply_step_or_start` function but with the `verify_authorization` specialized to either signature verification, SNARK verification, or nothing.
+
+
+Under this model, executing a transaction works as follows. Let `t = [t1; t2; t3]`
+
+### Fees and proofs required
+
+Instead of 2 proofs per transaction as is required now, we will switch to 2 proofs per "party".
+
+Similarly, we should switch to ordering transactions in the transaction pool by `fee / number of parties`.
+
+## Benefits of this approach
+
+The main benefit of this approach is that we can have a small number of base circuits, each of which has at most one verifier inside of it, while still supporting transactions containing arbitrary numbers of parties and proofs. This enables such applications as multi-token swaps and snapp interactions involving arbitrarily many accounts.
+
+Another benefit is the simplified and unified implementation for transaction application logic (both inside and outside the SNARK).
+
+## Eliminating all other user-command types
+
+Ideally, eventually, for simplicity, we will replace the implementation of transaction logic and transaction SNARK for the existing "special case" transactions (of payments and stake delegations) into sequences of "parties" as above. We can still keep the special case variants in the transaction type if desired.
+
+If we do this in the most straightforward way, a payment would go from occupying one leaf in the scan state to either 2 or 3 (if there is a separate fee payer). However, the proofs corresponding to these leaves would be correspondingly simpler. That said, there probably would be some efficiency loss and so if we want to avoid that, we can make circuits that "unroll the loop" and execute several parties per circuit.
+
+Specifically, for any sequence of authorization types, we can make a corresponding circuit to execute a sequence of parties with those authorization types. For example, it might be worth having a special circuit for the authorization type sequence `[ Signature; None ]` for a simple payment transaction that executes one party with a Signature authorization (the sender), and then one with no authorization (the receiver).
+
+## Potential issues
+
+- Backwards compatibility
+	+ Before changing the special case transactions into the above, we will have to make sure all signers are updated as the signed payload will change.
+- Transaction pool sorting
+	+ Currently, transactions in the transaction pool are sorted by sender by nonce. If general sequences of parties are allowed as transactions, this will not work and we will have to figure out another way to order things.
```

Comment:
It would be nice to explore some ideas how we would sort the transaction pool in this transaction model, at least so that we have confidence there is a reasonable mechanism for doing so. In particular, it seems that transaction sequencing logic could get really complex with this transaction model.